sdks/python/apache_beam/runners/direct/executor.py

```diff
@@ -291,45 +294,58 @@ def call(self):
 
     side_input_values = [self._side_input_values[side_input]
                          for side_input in self._applied_ptransform.side_inputs]
-
     try:
-      evaluator = self._transform_evaluator_registry.get_evaluator(
-          self._applied_ptransform, self._input_bundle,
-          side_input_values, scoped_metrics_container)
-
-      if self._fired_timers:
-        for timer_firing in self._fired_timers:
-          evaluator.process_timer_wrapper(timer_firing)
-
-      if self._input_bundle:
-        for value in self._input_bundle.get_elements_iterable():
-          evaluator.process_element(value)
+      self.attempt_call(metrics_container,
+                        scoped_metrics_container,
+                        side_input_values)
+    except Exception as e:
+      if self._retry_count == self._MAX_RETRY_PER_BUNDLE:
+        logging.error('Giving up after %s attempts.',
+                      self._MAX_RETRY_PER_BUNDLE)
+        self._completion_callback.handle_exception(self, e)
 
-      with scoped_metrics_container:
-        result = evaluator.finish_bundle()
-        result.logical_metric_updates = metrics_container.get_cumulative()
-
-      if self._evaluation_context.has_cache:
-        for uncommitted_bundle in result.uncommitted_output_bundles:
-          self._evaluation_context.append_to_cache(
-              self._applied_ptransform, uncommitted_bundle.tag,
-              uncommitted_bundle.get_elements_iterable())
-        undeclared_tag_values = result.undeclared_tag_values
-        if undeclared_tag_values:
-          for tag, value in undeclared_tag_values.iteritems():
-            self._evaluation_context.append_to_cache(
-                self._applied_ptransform, tag, value)
-
-      self._completion_callback.handle_result(self, self._input_bundle, result)
-      return result
-    except Exception as e:  # pylint: disable=broad-except
-      self._completion_callback.handle_exception(self, e)
     finally:
       self._evaluation_context.metrics().commit_physical(
           self._input_bundle,
           metrics_container.get_cumulative())
       self._transform_evaluation_state.complete(self)
 
+  @retry(stop_max_attempt_number=4)
```

Comment:
<!--new_thread; commit:eb2acfa855db99ab4b826344656ec29dcdb175e3; resolved:0-->
Please see above regarding usage of `@retry`.